Drop Result for ChannelMessageHandler methods

ariard opened the pull request:
Further work to simplify our error management by removing ChannelMessageHandler errors and instead sending back HandleError for the P2PHandler. 

If user, need more errors we should return APIErrors, but at least we should be have a clear distinction between errors for the user and errors sent to the peer.

TheBlueMatt commented:
I think I like this idea, pending making sure every callsite can actually support this, but why not just do it inside handle_error!()? It seems to me you can just do all the logic we have now, but instead of returning the LightningError, we can just push it as an event (as long as its the last thing that happens during the message processing).

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -429,19 +429,21 @@ pub struct ChannelDetails {
 
 macro_rules! handle_error {
-	($self: ident, $internal: expr) => {
+	($self: ident, $internal: expr, $their_node_id: expr, $locked_channel_state: expr) => {
```
Hmm, this does end up taking an extra lock in the happy case. Maybe split the macro into handle_error_locked and handle_error_unlocked?

ariard replied:
They were 2 callsites where we may don't have an error, modify a bit to avoid superfluous lock tacking if this is the case. Better than yet another error macro

TheBlueMatt replied:
Hmm? No, nearly ever message-handling callsite this is the case?

ariard replied:
Ah yeah corrected, it's a bit ugly rn but ofc I'm still planning a dry-up following PR

TheBlueMatt commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -516,6 +506,10 @@ pub fn do_test(data: &[u8]) {
 							// Can be generated due to a payment forward being rejected due to a
 							// channel having previously failed a monitor update
 						},
+						events::MessageSendEvent::HandleError { .. } => {
```
This really needs to check that the error in question is an IgnoreError action, and panic otherwise.

TheBlueMatt commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -532,6 +526,7 @@ pub fn do_test(data: &[u8]) {
 							events::MessageSendEvent::SendChannelReestablish { .. } => {},
 							events::MessageSendEvent::SendFundingLocked { .. } => {},
 							events::MessageSendEvent::PaymentFailureNetworkUpdate { .. } => {},
+							events::MessageSendEvent::HandleError { .. } => {},
```
Same here - this indicates exactly the type of failure this fuzz target exists to catch.

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -190,8 +190,8 @@ fn do_test_monitor_temporary_update_fail(disconnect_count: usize) {
-				if let Err(msgs::LightningError{err, action: msgs::ErrorAction::IgnoreError }) = nodes[0].node.handle_commitment_signed(&nodes[1].node.get_our_node_id(), commitment_signed) {
-					assert_eq!(err, "Previous monitor update failure prevented generation of RAA");
+				nodes[0].node.handle_commitment_signed(&nodes[1].node.get_our_node_id(), commitment_signed);
+				if let MessageSendEvent::HandleError { .. } = nodes[0].node.get_and_clear_pending_msg_events()[0] {
```
Can you restore the error-message-checking that was here previously?

TheBlueMatt replied:
Looks like this is still missing the IgnoreError check that was here? Also, why not check the error message inside the HandleError?

ariard commented:
Should be good now, modified Logger trait a bit to access logged lines and so keep test hardness as it is. Also added panic-if-not-IgnoreError in fuzz/src/chanmon_consistency.rs

TheBlueMatt commented:
Why do you need to extend the visibility of the ChannelManager::logger? Its a shared reference so you should be able to just keep the original copy and check against that, no?

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -515,8 +517,9 @@ fn test_monitor_update_fail_cs() {
-			if let msgs::LightningError { err, action: msgs::ErrorAction::IgnoreError } = nodes[0].node.handle_commitment_signed(&nodes[1].node.get_our_node_id(), &updates.commitment_signed).unwrap_err() {
-				assert_eq!(err, "Failed to update ChannelMonitor");
+			nodes[0].node.handle_commitment_signed(&nodes[1].node.get_our_node_id(), &updates.commitment_signed);
+			if let MessageSendEvent::HandleError { .. } = nodes[0].node.get_and_clear_pending_msg_events()[0] {
```
Same here as above - why not keep the IgnoreError and message-in-HandleError check that was here (even if you're also gonna add the log message tests)?

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1167,20 +1172,12 @@ impl<ChanSigner: ChannelKeys> ChannelManager<ChanSigner> {
-		match handle_error!(self, err) {
-			Ok(_) => unreachable!(),
-			Err(e) => {
-				if let msgs::ErrorAction::IgnoreError = e.action {
```
IIRC, that was quite intentional. Because you do want to log error even if it's a IgnoreError one but then was planning in further commit to match it out in the handle_error macro to avoid passing for nothing to peer_handler. I think that's good for now to pass it to peer_handler, it's just logged and ignored..

TheBlueMatt replied:
No? The IgnoreErrors that are generated today are all things where you didn't already get back an APIError directly (as the user). There is no need to send duplicative errors (though logging them is, of course, fine IMO).

ariard replied:
I think there is already duplicative errors right now in send_payment/update_fee for the non-IgnoreError cases? Inside  handle_error, you don't know if an APIError is going to be returned, so for not generating a HandleError when  we're returning an APIError anyway I need to filter out all IgnoreError, and so modify test in consequence again ?

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -1466,21 +1485,23 @@ fn test_monitor_update_on_pending_forwards() {
 	*nodes[1].chan_monitor.update_ret.lock().unwrap() = Err(ChannelMonitorUpdateErr::TemporaryFailure);
 	expect_pending_htlcs_forwardable!(nodes[1]);
 	check_added_monitors!(nodes[1], 1);
-	assert!(nodes[1].node.get_and_clear_pending_msg_events().is_empty());
```
lol you added an extra line to do the same thing :(. Also, there's a bunch of let msg_events = ... assert!(msg_events.len(), 0) that could be rewritten as just a .is_empty().

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -1595,23 +1618,19 @@ fn do_during_funding_monitor_fail(fail_on_generate: bool, restore_between_fails:
-			}
-		} else { panic!(); }
-
-		assert!(nodes[0].node.get_and_clear_pending_events().is_empty());
```
Hmm, seem to have lost this?

TheBlueMatt commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -491,8 +495,8 @@ macro_rules! commitment_signed_dance {
 			if $fail_backwards {
-				assert!($node_a.node.get_and_clear_pending_events().is_empty());
-				assert!($node_a.node.get_and_clear_pending_msg_events().is_empty());
+				//assert!($node_a.node.get_and_clear_pending_events().is_empty());
```
Looks like you didn't finish working on this?

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -952,17 +952,21 @@ fn do_test_shutdown_rebroadcast(recv_count: u8) {
-			let msgs::ErrorMessage {ref channel_id, ..} = msg;
-			assert_eq!(*channel_id, chan_1.2);
+		nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &node_0_2nd_reestablish);
+		if let MessageSendEvent::HandleError { ref action, .. } = nodes[1].node.get_and_clear_pending_msg_events()[0] {
```
Should check that there's only one msg event here.

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -1450,14 +1469,19 @@ fn do_channel_reserve_test(test_recv: bool) {
+			if let &MessageSendEvent::HandleError { ref action, .. } = handle_error {
+				match action {
+					&ErrorAction::SendErrorMessage { ref msg } => {
+						assert_eq!(msg.data, "Remote HTLC add would put them over their reserve value");
```
You can DRY this pattern up a ton (and fix a few cases where you forgot the trailing } else { panic!() }) by moving these matches into the macro.
